spec/requests/users/nbp_wallet/finalize_spec.rb

```diff
@@ -0,0 +1,218 @@
+      expect(User.order(:created_at).last).not_to be_confirmed
+    end
+
+    it 'asks the user to verify the email address' do
+      finalize_request
+      expect(response).to redirect_to home_index_path
+      expect(flash[:notice]).to include I18n.t('devise.registrations.signed_up_but_unconfirmed')
+    end
+  end
+
+  context 'with errors' do
+    shared_examples 'a handled erroneous request' do |error_message|
+      it 'does not create a user' do
+        expect { finalize_request }.not_to change(User, :count)
+      end
+
+      it 'creates no UserIdentities' do
+        expect { finalize_request }.not_to change(UserIdentity, :count)
+      end
+
+      it 'redirects to the connect page' do
+        finalize_request
+        expect(response).to redirect_to nbp_wallet_connect_users_path
+      end
+
+      it 'displays an error message' do
+        finalize_request
+        expect(flash[:alert]).to eq error_message
+      end
+
+      it 'does not send a confirmation mail' do
+        expect { finalize_request }.not_to change(ActionMailer::Base, :deliveries)
+      end
+    end
+
+    shared_examples 'a documented erroneous request' do |error|
+      it 'passes the error reason to Sentry' do
+        expect(Sentry).to receive(:capture_exception) do |e|
+          expect(e).to be_a error
+        end
+        finalize_request
+      end
+    end
+
+    context 'without the session' do
+      let(:session_params) { {} }
+
+      before do
+        finalize_request
+      end
+
+      it_behaves_like 'an unauthorized request'
+    end
+
+    context 'when the connector is down' do
+      before { allow(Enmeshed::Relationship).to receive(:pending_for).with(uid).and_raise(Faraday::ConnectionFailed) }
+
+      it 'redirects to the connect page' do
+        finalize_request
+        expect(response).to redirect_to(nbp_wallet_connect_users_path)
+      end
+    end
+
+    context 'when an attribute is missing' do
+      # `Enmeshed::ConnectorError` is unknown until 'lib/enmeshed/connector.rb' is loaded, because it's defined there
+      require 'enmeshed/connector'
+
+      let(:relationship) do
+        instance_double(Enmeshed::Relationship, accept!: false, reject!: true)
+      end
+
+      before do
+        allow(Enmeshed::Relationship).to receive(:pending_for).with(uid).and_return(relationship)
+        allow(relationship).to receive(:userdata).and_raise(Enmeshed::ConnectorError, 'EMailAddress must not be empty')
+      end
+
+      it_behaves_like 'a handled erroneous request', I18n.t('common.errors.generic')
+      it_behaves_like 'a documented erroneous request', Enmeshed::ConnectorError
+    end
+
+    context 'with an invalid status group' do
+      let(:relationship) do
+        instance_double(Enmeshed::Relationship,
+          reject!: true,
+          userdata: {
+            email: 'john.oliver@example103.org',
+            first_name: 'john',
+            last_name: 'oliver',
+            status_group: nil,
```

Comment:
`nil` is invalid, but another free text entry might be invalid, too. Since you are currently checking for the presence of `Enmeshed::RelationshipTemplate::REQUIRED_ATTRIBUTES`, I would actually prefer to test for another present, but invalid string here. A non-present string is actually checked [here](https://github.com/openHPI/codeharbor/blob/aa8fcbfb55f11c0cf4aead99f41eee5dea30a998/spec/lib/enmeshed/relationship_spec.rb#L108) already.

Reply:
The validations in  `Enmeshed::Relationship` prevent invalid free texts and only pass on one of these three values on: `:learner`, `:educator` and `nil`.

I'd rather add a test with an invalid free text entry to the specs of `Enmeshed::Relationship` to make sure it transforms it to `nil`. And this test here makes sure that controller handles it correctly with a helpful error message for the user.